File: packages/ruby/lib/menuella/food_safety.rb
# frozen_string_literal: true

require "json"
require "monitor"

require_relative "food_safety/version"

module Menuella
  # The Menuella food-safety vocabulary: EU Reg. 1169/2011 Annex II allergens
  # and the Menuella declarations, in six languages.
  #
  # Semantic keys instead of country-specific numbers — store the key, render
  # the code, never the reverse.
  #
  #   de = Menuella::FoodSafety.disclosures("de")
  #   de.allergens.find { _1.key == "WHEAT" }.declaration
  #   # => "Enthält Getreide und glutenhaltige Erzeugnisse"
  #
  # This module does *no* i18n. Hand it the locale your application already
  # resolved; it will not sniff the environment, negotiate, or quietly fall
  # back.
  #
  # Everything it returns is frozen. The dataset is a shared constant, and a
  # caller that could mutate one bundle would be mutating every other caller's
  # copy in the same process.
  module FoodSafety
    # Base class for everything this module raises, so a caller can rescue the
    # whole surface without naming each one.
    class Error < StandardError; end

    # The locale has no bundle. FoodSafety.locales lists the ones that do.
    class UnsupportedLocaleError < Error
      attr_reader :locale

      def initialize(locale)
        @locale = locale
        super("no disclosures for locale #{locale.inspect} " \
              "(available: #{FoodSafety.locales.join(', ')})")
      end
    end

    # The name has no glyph. FoodSafety.icon_names lists the ones that do.
    class UnknownIconError < Error
      attr_reader :name

      def initialize(name)
        @name = name
        super("no icon named #{name.inspect}")
      end
    end

    # One allergen, resolved for a locale.
    #
    # +key+ is what a product row stores. +group+ is its LMIV group; +CEREALS+
    # and +TREE_NUTS+ are groups but never keys, because the law requires
    # naming the specific grain or nut. +declaration+ is the sentence with
    # legal force — the one that must reach the guest.
    Allergen = Data.define(:key, :group, :is_member, :icon, :name, :declaration, :description) do
      # Ruby spells a predicate with a question mark; the dataset spells it
      # +isMember+. Both names reach the same field.
      alias_method :member?, :is_member
    end

    # One declaration — an additive, beverage or product note — for a locale.
    Declaration = Data.define(:key, :category, :icon, :name, :description)

    # Every disclosure for one locale, ready to render.
    Disclosures = Data.define(:locale, :fallback_locale, :allergens, :declarations)

    # The footnote-code scheme: letters for allergens, numbers for declarations.
    #
    # Codes are what a printed menu shows. Keys are what a database stores, and
    # the projection only runs in that direction — a code is a rendering detail
    # that varies by region and template, a key is not.
    Codes = Data.define(:scheme, :allergens, :declarations)

    # A single shape in a glyph. +attributes+ uses SVG spelling
    # (<tt>fill-rule</tt>, not <tt>fillRule</tt>) and is sorted by name, so
    # rendered markup is byte-stable.
    IconNode = Data.define(:tag, :attributes)

    # A glyph as data, for callers that build shapes rather than markup.
    Icon = Data.define(:view_box, :nodes)

    DATA_DIR = File.join(__dir__, "food_safety", "data").freeze
    private_constant :DATA_DIR

    # The JSON carries React attribute spelling, because its first consumer is
    # React. Markup needs the SVG one — and an unmapped name still draws, just
    # without the even-odd rule, so the bug would be a subtly wrong glyph
    # rather than a missing one.
    SVG_ATTRIBUTE = { "fillRule" => "fill-rule", "clipRule" => "clip-rule" }.freeze
    private_constant :SVG_ATTRIBUTE

    ESCAPES = { "&" => "&amp;", "<" => "&lt;", ">" => "&gt;", '"' => "&quot;", "'" => "&#39;" }.freeze
    private_constant :ESCAPES

    # A Monitor, not a Mutex: #allergen_keys memoises a value it computes by
    # calling #disclosures, which memoises too. A Mutex is not reentrant, so
    # that nesting would deadlock on the first call rather than on some rare
    # race — the kind of bug that passes every test written after it.
    LOCK = Monitor.new
    private_constant :LOCK

    class << self
      # Locales with a prebuilt bundle.
      def locales
        @locales ||= %w[de en es fr it tr].freeze
      end

      # The locale a bundle falls back to for anything it does not itself carry.
      def fallback_locale
        "en"
      end

      # Every disclosure for +locale+, ready to render.
      #
      # Raises UnsupportedLocaleError rather than falling back: a silently wrong
      # language on an allergen panel is worse than a loud failure. The caller
      # knows which locales it supports, and #locale? is there to ask.
      def disclosures(locale)
        raise UnsupportedLocaleError, locale unless locale?(locale)

        memo(:bundles, locale) do
          raw = read("bundles/#{locale}")
          Disclosures.new(
            locale: raw.fetch("locale").freeze,
            fallback_locale: raw.fetch("fallbackLocale").freeze,
            allergens: raw.fetch("allergens").map { |a|
              Allergen.new(
                key: a.fetch("key").freeze,
                group: a.fetch("group").freeze,
                is_member: a.fetch("isMember"),
                icon: a.fetch("icon").freeze,
                name: a.fetch("name").freeze,
                declaration: a.fetch("declaration").freeze,
                description: a.fetch("description").freeze,
              )
            }.freeze,
            declarations: raw.fetch("declarations").map { |d|
              Declaration.new(
                key: d.fetch("key").freeze,
                category: d.fetch("category").freeze,
                icon: d.fetch("icon").freeze,
                name: d.fetch("name").freeze,
                description: d.fetch("description").freeze,
              )
            }.freeze,
          )
        end
      end

      # The footnote-code scheme.
      def codes
        memo(:codes) do
          raw = read("codes")
          Codes.new(
            scheme: raw.fetch("scheme").freeze,
            allergens: deep_freeze(raw.fetch("allergens")),
            declarations: deep_freeze(raw.fetch("declarations")),
          )
        end
      end

      # The scheme #codes belong to.
      def code_scheme
        codes.scheme
      end

      # Every selectable allergen key, in canonical order.
      #
      # Derived from the dataset, never hand-listed. Keys are
      # locale-independent, so this reads the fallback bundle rather than
      # taking a locale argument.
      def allergen_keys
        memo(:allergen_keys) { disclosures(fallback_locale).allergens.map(&:key).freeze }
      end

      # Every declaration key, in canonical order.
      def declaration_keys
        memo(:declaration_keys) { disclosures(fallback_locale).declarations.map(&:key).freeze }
      end

      # Every icon name that has a glyph, sorted.
      def icon_names
        memo(:icon_names) { icons.keys.sort.freeze }
      end

      # True when +value+ is a locale with a bundle.
      def locale?(value)
        locales.include?(value)
      end

      # True when +value+ is a current allergen key. Retired keys and
      # display-only groups return false.
      def allergen_key?(value)
        allergen_keys.include?(value)
      end

      # True when +value+ is a current declaration key.
      def declaration_key?(value)
        declaration_keys.include?(value)
      end

      # The glyph named +name+, as data.
      #
      # Every shape paints with +currentColor+, so a glyph inherits the
      # surrounding text colour and follows a light or dark theme with no
      # second asset.
      def icon(name)
        icons.fetch(name) { raise UnknownIconError, name }
      end

      # The glyph named +name+ as an <tt><svg></tt> string, for anything that
      # interpolates markup — server-rendered HTML, e-mail, PDF.
      #
      # Decorative by default: emits +aria-hidden+ unless +title:+ is given,
      # which switches it to <tt>role="img"</tt> with a <tt><title></tt>. These
      # glyphs carry legal meaning, so render one *alongside* its declaration
      # text, never instead of it — the safe default is the free one.
      def icon_to_svg(name, size: 24, css_class: nil, title: nil)
        glyph = icon(name)

        out = +%(<svg xmlns="http://www.w3.org/2000/svg" viewBox="#{escape(glyph.view_box)}")
        out << %( width="#{size.to_i}" height="#{size.to_i}" fill="none")
        out << %( class="#{escape(css_class)}") if css_class && !css_class.empty?

        out << if title && !title.empty?
                 %( role="img"><title>#{escape(title)}</title>)
               else
                 %( aria-hidden="true" focusable="false">)
               end

        glyph.nodes.each do |node|
          out << "<#{node.tag}"
          node.attributes.each { |key, value| out << %( #{key}="#{escape(value)}") }
          out << "/>"
        end

        out << "</svg>"
        out.freeze
      end

      # The raw packaged JSON for +name+, for tooling that wants the dataset
      # rather than the objects. Returns a fresh string each call.
      def dataset(name)
        File.read(File.join(DATA_DIR, "#{name.sub(/\.json\z/, '')}.json"), encoding: "UTF-8")
      end

      private

      def icons
        memo(:icons) do
          read("icons").to_h { |name, glyph|
            [
              name.freeze,
              Icon.new(
                view_box: glyph.fetch("viewBox").freeze,
                nodes: glyph.fetch("nodes").map { |tag, attributes|
                  IconNode.new(
                    tag: tag.freeze,
                    # Sorted so the markup is deterministic. Ruby preserves Hash
                    # insertion order, so this would happen to be stable — but
                    # it would then depend on the order the generator wrote the
                    # JSON in, which is not a contract anyone stated.
                    attributes: attributes
                      .to_h { |k, v| [(SVG_ATTRIBUTE[k] || k).freeze, v.freeze] }
                      .sort.to_h.freeze,
                  )
                }.freeze,
              ),
            ]
          }.freeze
        end
      end

      def read(name)
        JSON.parse(File.read(File.join(DATA_DIR, "#{name}.json"), encoding: "UTF-8"))
      end

      # Memoised behind a lock. Without it two threads racing on first call
      # would each parse the file and one would win — harmless here, but only
      # by luck, and #disclosures is exactly the kind of thing a web server
      # calls from every request thread at once.
      def memo(bucket, key = nil)
        LOCK.synchronize do
          store = (@memo ||= {})[bucket] ||= {}
          store.fetch(key) { store[key] = yield }
        end
      end

      def deep_freeze(hash)
        hash.to_h { |k, v| [k.freeze, v.freeze] }.freeze
      end

      # The path data is ours, but +icon_to_svg+'s title and class are
      # caller-supplied and land in markup.
      def escape(value)
        value.to_s.gsub(/[&<>"']/, ESCAPES)
      end
    end
  end
end

File: packages/swift/Sources/MenuellaFoodSafety/FoodSafety.swift
import Foundation

/// Everything that can go wrong reading the packaged dataset.
public enum FoodSafetyError: Error, LocalizedError, Sendable, Equatable {
    /// The locale has no bundle. ``FoodSafety/locales`` lists the ones that do.
    case unsupportedLocale(String)

    /// A resource is missing, which means the package was built without its data.
    case missingResource(String)

    /// A resource is present but could not be decoded.
    case invalidResource(String, underlying: String)

    /// The name has no glyph.
    case unknownIcon(String)

    public var errorDescription: String? {
        switch self {
        case let .unsupportedLocale(locale):
            return "No disclosures for locale \"\(locale)\". "
                + "Available: \(FoodSafety.locales.joined(separator: ", "))."
        case let .missingResource(resource):
            return "Packaged dataset is missing \(resource)."
        case let .invalidResource(resource, underlying):
            return "Could not decode \(resource): \(underlying)"
        case let .unknownIcon(name):
            return "No icon named \"\(name)\"."
        }
    }
}

/// The Menuella food-safety vocabulary: EU Reg. 1169/2011 Annex II allergens
/// and the Menuella declarations, in six languages.
///
/// Semantic keys instead of country-specific numbers — store the key, render
/// the code, never the reverse.
///
/// ```swift
/// let de = try FoodSafety.disclosures(locale: "de")
/// de.allergens.first { $0.key == "WHEAT" }?.declaration
/// // "Enthält Getreide und glutenhaltige Erzeugnisse"
/// ```
///
/// This type does **not** do i18n. Hand it the locale your app already
/// resolved; it will not sniff, negotiate, or quietly fall back.
public enum FoodSafety {

    /// Locales with a prebuilt bundle.
    public static let locales: [String] = ["de", "en", "es", "fr", "it", "tr"]

    /// The locale a bundle falls back to for anything it does not itself carry.
    public static let fallbackLocale = "en"

    // Decoded bundles are cached behind a lock rather than recomputed. A
    // `static var` would not be Sendable under strict concurrency, and an actor
    // would force every call site to be async for what is a pure lookup.
    private static let cache = Cache()

    private final class Cache: @unchecked Sendable {
        private let lock = NSLock()
        private var bundles: [String: Disclosures] = [:]
        private var icons: [String: Icon]?
        private var codes: Codes?

        func bundle(_ locale: String, _ make: () throws -> Disclosures) rethrows -> Disclosures {
            lock.lock()
            defer { lock.unlock() }
            if let hit = bundles[locale] { return hit }
            let made = try make()
            bundles[locale] = made
            return made
        }

        func allIcons(_ make: () throws -> [String: Icon]) rethrows -> [String: Icon] {
            lock.lock()
            defer { lock.unlock() }
            if let hit = icons { return hit }
            let made = try make()
            icons = made
            return made
        }

        func allCodes(_ make: () throws -> Codes) rethrows -> Codes {
            lock.lock()
            defer { lock.unlock() }
            if let hit = codes { return hit }
            let made = try make()
            codes = made
            return made
        }
    }

    // MARK: - Disclosures

    /// Every disclosure for `locale`, ready to render.
    ///
    /// Throws rather than falling back: a silently wrong language on an
    /// allergen panel is worse than a loud failure. The caller knows which
    /// locales it supports, and ``isLocale(_:)`` is there to ask.
    public static func disclosures(locale: String) throws -> Disclosures {
        guard isLocale(locale) else {
            throw FoodSafetyError.unsupportedLocale(locale)
        }
        return try cache.bundle(locale) {
            try decode(Disclosures.self, named: locale, in: "Data/bundles")
        }
    }

    /// The footnote-code scheme: letters for allergens, numbers for declarations.
    public static func codes() throws -> Codes {
        try cache.allCodes { try decode(Codes.self, named: "codes", in: "Data") }
    }

    /// The scheme these codes belong to.
    public static func codeScheme() throws -> String { try codes().scheme }

    // MARK: - Vocabulary

    /// Every selectable allergen key, in canonical order.
    ///
    /// Derived from the dataset, never hand-listed. Keys are locale-independent,
    /// so this reads the fallback bundle rather than taking a locale argument.
    public static func allergenKeys() throws -> [String] {
        try disclosures(locale: fallbackLocale).allergens.map(\.key)
    }

    /// Every declaration key, in canonical order.
    public static func declarationKeys() throws -> [String] {
        try disclosures(locale: fallbackLocale).declarations.map(\.key)
    }

    /// Every icon name that has a glyph.
    public static func iconNames() throws -> [String] {
        try allIcons().keys.sorted()
    }

    /// True when `value` is a locale with a bundle.
    public static func isLocale(_ value: String) -> Bool { locales.contains(value) }

    /// True when `value` is a current allergen key. Retired keys return false.
    public static func isAllergenKey(_ value: String) throws -> Bool {
        try allergenKeys().contains(value)
    }

    /// True when `value` is a current declaration key.
    public static func isDeclarationKey(_ value: String) throws -> Bool {
        try declarationKeys().contains(value)
    }

    // MARK: - Icons

    /// The glyph named `name`, as data.
    ///
    /// Every shape paints with `currentColor`, so a glyph inherits the
    /// surrounding text colour and follows light and dark with no second asset.
    public static func icon(named name: String) throws -> Icon {
        guard let icon = try allIcons()[name] else {
            throw FoodSafetyError.unknownIcon(name)
        }
        return icon
    }

    /// The glyph named `name` as an `<svg>` string, for anything that
    /// interpolates markup — server-rendered HTML, e-mail, PDF.
    ///
    /// Decorative by default: emits `aria-hidden` unless `title` is given,
    /// which switches it to `role="img"` with a `<title>`. These glyphs carry
    /// legal meaning, so render one *alongside* its declaration text, never
    /// instead of it — the safe default is the free one.
    public static func iconToSVG(
        named name: String,
        size: Int = 24,
        cssClass: String? = nil,
        title: String? = nil
    ) throws -> String {
        let icon = try icon(named: name)
        var out = #"<svg xmlns="http://www.w3.org/2000/svg" viewBox=""#
        out += escape(icon.viewBox)
        out += "\" width=\"\(size)\" height=\"\(size)\" fill=\"none\""

        if let cssClass, !cssClass.isEmpty {
            out += " class=\"\(escape(cssClass))\""
        }

        if let title, !title.isEmpty {
            out += " role=\"img\"><title>\(escape(title))</title>"
        } else {
            out += " aria-hidden=\"true\" focusable=\"false\">"
        }

        for node in icon.nodes {
            out += "<\(node.tag)"
            // Sorted so the output is deterministic — a Swift Dictionary has no
            // stable order, and unstable markup breaks snapshot tests and caches.
            for key in node.attributes.keys.sorted() {
                out += " \(key)=\"\(escape(node.attributes[key]!))\""
            }
            out += "/>"
        }

        return out + "</svg>"
    }

    // MARK: - Loading

    // The JSON carries React attribute spelling, because its first consumer is
    // React. Markup needs the SVG one — and an unmapped name still draws, just
    // without the even-odd rule, so the bug would be a subtly wrong glyph
    // rather than a missing one.
    private static let svgAttribute = ["fillRule": "fill-rule", "clipRule": "clip-rule"]

    private static func allIcons() throws -> [String: Icon] {
        try cache.allIcons {
            // Decoded by hand rather than through Codable: a node is
            // ["path", { … }] — a heterogeneous array whose first element is a
            // string and second an object — which Codable models badly.
            let raw = try decode([String: RawIcon].self, named: "icons", in: "Data")
            return raw.mapValues { icon in
                Icon(
                    viewBox: icon.viewBox,
                    nodes: icon.nodes.map { node in
                        IconNode(
                            tag: node.tag,
                            attributes: Dictionary(
                                uniqueKeysWithValues: node.attributes.map {
                                    (svgAttribute[$0.key] ?? $0.key, $0.value)
                                }
                            )
                        )
                    }
                )
            }
        }
    }

    private struct RawIcon: Decodable {
        let viewBox: String
        let nodes: [RawNode]

        struct RawNode: Decodable {
            let tag: String
            let attributes: [String: String]

            init(from decoder: any Decoder) throws {
                var container = try decoder.unkeyedContainer()
                tag = try container.decode(String.self)
                attributes = try container.decode([String: String].self)
            }
        }
    }

    private static func decode<T: Decodable>(
        _ type: T.Type,
        named name: String,
        in subdirectory: String
    ) throws -> T {
        guard let url = Bundle.module.url(
            forResource: name,
            withExtension: "json",
            subdirectory: subdirectory
        ) else {
            throw FoodSafetyError.missingResource("\(subdirectory)/\(name).json")
        }

        do {
            return try JSONDecoder().decode(type, from: Data(contentsOf: url))
        } catch {
            throw FoodSafetyError.invalidResource(
                "\(name).json",
                underlying: String(describing: error)
            )
        }
    }

    /// The path data is ours, but `iconToSVG`'s title is caller-supplied.
    private static func escape(_ value: String) -> String {
        value
            .replacingOccurrences(of: "&", with: "&amp;")
            .replacingOccurrences(of: "<", with: "&lt;")
            .replacingOccurrences(of: ">", with: "&gt;")
            .replacingOccurrences(of: "\"", with: "&quot;")
            .replacingOccurrences(of: "'", with: "&#39;")
    }
}
